src/ray/raylet/scheduling_queue.cc

```diff
@@ -73,111 +76,107 @@ namespace ray {
-    // in the ready queue.
-    RecomputeMinResources();
-  }
   return true;
 }
 
-bool SchedulingQueue::TaskQueue::HasTask(const TaskID &task_id) const {
+bool TaskQueue::HasTask(const TaskID &task_id) const {
   return task_map_.find(task_id) != task_map_.end();
 }
 
-const ResourceSet &SchedulingQueue::TaskQueue::GetCurrentResourceLoad() const {
-  return ready_tasks_metadata_.GetCurrentResourceLoad();
+// const ResourceSet &TaskQueue::GetCurrentResourceLoad() const {
+//   return ready_tasks_metadata_.GetCurrentResourceLoad();
+// }
+
+const std::list<Task> &TaskQueue::GetTasks() const { return task_list_; }
+
+bool ReadyQueue::AppendTask(const TaskID &task_id, const Task &task) {
+  bool result = TaskQueue::AppendTask(task_id, task);
+  // Resource bookkeeping
+  const auto &task_resources = task.GetTaskSpecification().GetRequiredResources();
+  current_resource_load_.AddResources(task_resources);
+  UpdateMinOnAdd(task_resources);
+  return result;
 }
 
-void SchedulingQueue::TaskQueue::RecomputeMinResources() {
-  for (const auto &task : task_list_) {
-    ready_tasks_metadata_.UpdateMinOnAdd(task.GetTaskSpecification().GetRequiredResources());
-  };
+bool ReadyQueue::RemoveTask(const TaskID &task_id,
+                            std::vector<Task> *removed_tasks) {
+  auto task_found_iterator = task_map_.find(task_id);
+  if (task_found_iterator != task_map_.end()) {
+    const auto &task_resources =
+      task_found_iterator->second->GetTaskSpecification().GetRequiredResources();
+      // Update sum resources.
+      current_resource_load_.SubtractResourcesStrict(task_resources);
+      // Update min resources.
+      UpdateMinOnRemove(task_resources);
+  }
+  bool result = TaskQueue::RemoveTask(task_id, removed_tasks);
+  if (min_task_count_ <= 0) {
+    // Need to recompute the minimum resources required by any task
+    // in the ready queue.
+    RecomputeMinResources();
+  }
+  return result;
+}
+
+void ReadyQueue::UpdateMinOnAdd(const ResourceSet &resources) {
+  if (min_task_count_ <= 0) {
+    min_task_resources_ = resources;
+    min_task_count_= 1;
+    return;
+  }
+  if (resources.IsSubset(min_task_resources_)) {
+    if (min_task_resources_.IsSubset(resources)) {
+      // In this case, resources == min_task_resources_.
+      min_task_count_++;
+    } else {
+      min_task_resources_ = resources;
+      min_task_count_= 1;
+    }
+  }
 };
 
-const bool SchedulingQueue::TaskQueue::CanScheduleMinTask(const ResourceIdSet &local_available_resources) const {
-  RAY_CHECK(ready_tasks_metadata_.GetMinTaskCount() > 0);
-  return local_available_resources.Contains(ready_tasks_metadata_.GetMinTaskResources());
+void ReadyQueue::UpdateMinOnRemove(const ResourceSet &resources) {
+  if (min_task_resources_ == resources) {
+    min_task_count_--;
+  }
 };
 
+void ReadyQueue::RecomputeMinResources() {
+  for (const auto &task : task_list_) {
+    UpdateMinOnAdd(task.GetTaskSpecification().GetRequiredResources());
```

Comment:
we can speed this up to be O(logn) by maintaining a min heap

Reply:
Yeah, good idea! I'll leave it as a separate PR for now.